src/Data/TypeRepMap/Internal.hs

```diff
@@ -390,14 +387,19 @@ fromTriples kvs = TypeRepMap (GHC.fromList fpAs) (GHC.fromList fpBs) (GHC.fromLi
 ----------------------------------------------------------------------------
 
 fromSortedList :: forall a . [a] -> [a]
-fromSortedList l = IM.elems $ fst $ go 0 0 mempty (IM.fromList $ zip [0..] l)
+fromSortedList l = runST $ do
+    let n = length l
+    let arrOrigin = fromList l
```

Comment:
Since you've precalculated length of the array, you can use `fromListN` function to create `arrOrigin`. This will be more efficient:

* https://hackage.haskell.org/package/base-4.11.1.0/docs/GHC-Exts.html#v:fromListN